### DISCUSSION TITLE
Factories / Seeders in Service deprecated?

### DISCUSSION THREAD

**Discussion by ghost** [lucidarch/lucid#31] [Q&A] [answered]
Is it deprecated to make factories and seeders in services?
There's no command for making them for a service and since factories and seeders have namespaces now, it would require registering them in the service's provider which is ALSO not being done.

I feel like it should be properly decided and updated whether services should have separate database folders or share the laravel project's and maybe just have folders there.

**Comment by Batisska** [comment]
I am currently using standard seeds and factories. 
It seems to me that the standard ones are convenient. Since the models in the domains are still the same and will also be used

**Comment by Mulkave** [comment, marked as answer]
@mwkohler  if you had generated a service with `lucid make service {name}` there should be a `database` folder generated automatically with the following structure (assuming the service name is `Chat`):

```
app/Services/Chat/database
├── factories
├── migrations
└── seeds
```

----

> **In the examples below we assume the service name is `Chat`. Make sure you replace it with your service's name instead.**

### Factories

1. Update composer to autoload service's factories
	- Under  `autoload.psr-4` add `"App\\Services\\Chat\\Database\\Factories\\": "app/Services/Chat/database/factories/"`
         So your `composer.json` should have a similar section as this:
         ```json
          "autoload": {
                "psr-4": {
                    "App\\": "app/",
                    "Database\\Factories\\": "database/factories/",
                    "Database\\Seeders\\": "database/seeders/",
                    "App\\Services\\Chat\\Database\\Factories\\": "app/Services/Chat/database/factories/"
                }
        }
        ```
	- Run `composer dump-autoload`
2. Create factory class (e.g.  `php artisan make:factory --model=Message`)
3. Move factory class from `database/factories` to `app/Services/Chat/database/factories`
4. Change class namespace to `App\Services\Chat\Database\Factories`
5. Add `newFactory()` method to model. This method is an override that Laravel looks at before loading the default factory (see `HasFactory` trait)
```php
use App\Services\Chat\Database\Factories\MessageFactory;

protected static function newFactory()
{
    return app(MessageFactory::class);
}
```
6. Use the factory from the model as you would by default: `Message::factory()->make()`

-----

### Seeders

1. Update composer to autoload service's seeders
	- in `autoloat.psr-4` add `"App\\Services\\Chat\\Database\\Seeders\\": "app/Services/Chat/database/seeders/"`
        So your `composer.json` should have a similar section as this:
         ```json
          "autoload": {
                "psr-4": {
                    "App\\": "app/",
                    "Database\\Factories\\": "database/factories/",
                    "Database\\Seeders\\": "database/seeders/",
                    "App\\Services\\Chat\\Database\\Factories\\": "app/Services/Chat/database/factories/",
                    "App\\Services\\Chat\\Database\\Seeders\\": "app/Services/Chat/database/seeders/"
                }
        }
        ```
	- Run `composer dump-autoload`
2. Create seeder class (e.g. `php artisan make:seeder ChatSeeder`)
3. Rename directory `app/Services/Chat/database/seeds` to `app/Services/Chat/database/seeders` for consistency (will be default in Lucid 1.0.5)
4. Move seeder class to `app/Services/Chat/database/seeders`
5. Change class namespace to `App\Services\Chat\Database\Seeders`
6. Call `ChatSeeder::class` from `database/seeders/DatabaseSeeder.php`
```php
use App\Services\Chat\Database\Seeders\ChatSeeder;

public function run()
    {
        $this->call([
            ChatSeeder::class,
        ]);
    }
```
7. Run `php artisan db:seed`

---

Thank you @mwkohler for raising this topic that's certainly important and missing from the docs, I will be adding these instructions to the docs under a "Database" section.

**Reply by Mulkave** [reply]
This has now been added to the [docs](https://docs.lucidarch.dev/database/#factories). Next step is to create commands to automate some these steps such as `lucid make:factory` and `lucid make:seeder`.
